### ISSUE TITLE
Regression with defaulted Callable in dataclass.

### ISSUE THREAD

**Issue by adampauls** [microsoft/pyright#9222]
**Describe the bug**
There seems to be a regression between 1.1.383 and 1.1.384. 

```python
from collections.abc import Callable
from dataclasses import dataclass


@dataclass(frozen=True)
class Foo[T]:
    result_type: type[T]
    condition: Callable[[T], bool] = lambda _: True


class Bar:
    pass


Foo(result_type=Bar) # error:
# error: Argument of type "(_: T@Foo) -> Literal[True]" cannot be assigned to parameter "condition" of type "(T@Foo) -> bool" in function "__init__"
#    Type "(_: T@Foo) -> Literal[True]" is not assignable to type "(T@Foo) -> bool"
#      Parameter 1: type "T@Foo" is incompatible with type "T@Foo"
#        Type "Bar" is not assignable to type "T@Foo" (reportArgumentType)


```
This code was fine in 1.1.383. 

Note that the error does not appear if we do not use a dataclass:

```python
from collections.abc import Callable


class Foo[T]:
    def __init__(
        self,
        result_type: type[T],
        condition: Callable[[T], bool] = lambda _: True,
    ) -> None:
        pass


class Bar:
    pass


Foo(result_type=Bar) # no error
```

I know the semantics of the type variable are a little different between the two, but I believe the first one should type check just fine. 

**VS Code extension or command-line**
Command-line.

**Comment by adampauls** [comment]
Oh, nevermind, I need to use a default_factory here.

**Comment by erictraut** [comment]
I agree this is a bug. 

This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.385.
